### PR TITLE
Enable JaCoCo code coverage

### PR DESCRIPTION
**Pull request by basil**
Experimenting with JaCoCo code coverage.

### CODE REVIEW

**Comment by basil** [comment]
It is helpful to know that line coverage is at 41.73% and branch coverage is at 49.64%, but this isn't the information I really care about as a core maintainer. What I really care about is knowing, for a given PR, whether the changed lines were exercised by automated testing. That way I can be confident that the code was properly tested before I approve the PR. But my experience as an operator is that preserving this information can be a challenge in large repositories. In this repository there are 67.8 MiB of Java sources, spread out over 1,757 files. This can be a challenge depending on where and how the Jenkins controller stores build records and artifacts.

I can think of a few possible options offhand:

- Use something like `publishCoverage […] sourceFileResolver: sourceFiles('STORE_LAST_BUILD')` to store the source files for the last build in the build directory, where they could then be displayed natively in the Jenkins UI. I think this would provide the best UI, but it would come at the cost of an additional 68 MiB for each active branch (including PR branches), spread out over 1,757 files each. I am not sure if that is an appropriate usage pattern for our deployment.
- Use the [HTML Publisher](https://plugins.jenkins.io/htmlpublisher/) to publish the HTML files from the `coverage/target/site/jacoco-aggregate` directory. This directory is about 35 MiB for each build (less than the first option) spread out over 3,614 tiny files (greater than the first option). We'd probably want to use the "Keep past HTML reports" to only keep the HTML files from the last stable build of each active branch. This option has a slightly worse UI (not integrated as well with Jenkins). I am also not sure if this is an appropriate usage pattern for our deployment.
- Zip the HTML files from the `coverage/target/site/jacoco-aggregate` and archive the resulting zipped artifact. The resulting ZIP file is about 8 MiB and is a single file, which is nice. But the UI of having to download the ZIP file and extract it to view coverage information isn't great. Still, it's better than not having the data. Also, the `archiveArtifacts` step doesn't appear to have a way to discard the artifacts from previous builds of each active branch. This might be an option if the first two options are not appropriate for our deployment.

I'll have to spend some time thinking about which of these options (if any) is best for us, possibly doing some more research into how `ci.jenkins.io` is deployed.

**Comment by timja** [comment]
@uhafner has been rewriting how this is done in the code-coverage API plugin: https://github.com/jenkinsci/code-coverage-api-plugin/pull/265

Looks like it could be useful here, especially this bit of it `ZIP individual sources as well and decompress on access only`

**Comment by uhafner** [comment]
A student of mine currently also adds a delta coverage model on top of the current model in his master theses. The idea is to store and analyze only those elements that actually have been changed between the current build and the reference build. This is still work in progress (the computation of the Git delta is done) but will be available in the near future. 

Another option would be to publish the results to https://codecov.io? In jenkinsci/code-coverage-api-plugin#265 you can see how such a report looks like (comment "Codecov Report" from the coverage bot).

**Comment by basil** [comment]
While I originally had greater ambitions for this PR (to be able to show a coverage report for Java source files in CI), I never found the time to work on that. In the meantime, this PR represents an improvement over the status quo in two regards:

- It adds to CI an overall coverage report giving percentages of various types of coverage. For example, between the time I began working on this PR in October 2021 and the present, line coverage has increased from 41.73% to 42.09%.
- It adds the ability to get a detailed coverage report when running locally, including highlighting the source file of the code under test with the lines that got exercised by the test. While I would much rather have this in CI for ease of use when reviewing PRs, at least being able to run this locally gives me the information I need (albeit at the requirement of a local checkout and build). If I am already doing a local checkout and build to manually test a PR, I might as well also be able to view the coverage report from any existing test automation to see where the gaps are.

I cannot think of any downsides to merging this, and the upsides (while not as ambitious as originally planned) are non-negligible, so I am marking this PR as ready for review..

**Review by NotMyFault — APPROVED** [review]

File: test/pom.xml
Comment (timja) [inline test/pom.xml:356]:
```suggestion
                    <include>jenkins/**/*</include>
                    <include>org/jenkins/**/*</include>
```

**Review by timja — APPROVED** [review]
Possibly add org/jenkins as well?

File: test/pom.xml
Comment (basil) [reply test/pom.xml:356]: Done in commit 21e286c21f

**Review by timja — APPROVED** [review]

**Review by res0nance — APPROVED** [review]

**Comment by csmith131314** [comment]
Hi

**Comment by timja** [comment]
Breaks incrementals publishing:

```
[2022-04-19T10:14:37.420Z] Error: ZIP error: Error: No permissions for org/jenkins-ci/main/jenkins-coverage/2.344-rc32342.0b_e118fe44a_d/jenkins-coverage-2.344-rc32342.0b_e118fe44a_d.jar from https://ci.jenkins.io/job/Core/job/jenkins/job/PR-6408/44/artifact/**/*0b*e118fe44a*d*/*0b*e118fe44a*d*/*zip*/archive.zip
```

Noticed in https://github.com/jenkinsci/jenkins/pull/6408#discussion_r852906710

**Pull request merged**
